internal/resource/buffer.go

```diff
@@ -0,0 +1,110 @@
+package resource
+
+import (
+	"errors"
+	"io"
+)
+
+var (
+	ErrOutOfMemory    = errors.New("unable to grow buffer")
+	ErrOffsetTooLarge = errors.New("cannot seek past end of buffer")
+	ErrUnknownWhence  = errors.New("unknown whence")
+)
+
+type buffer struct {
+	buf    []byte
+	cursor int64
+	end    int64
+	cap    int64
+}
+
+func NewBuffer() *buffer {
+	return &buffer{}
+}
+
+func (b *buffer) grow(n int64) error {
+	newCap := 2*b.cap + n
+	buf, err := makeSlice(newCap)
+	if err != nil {
+		return err
+	}
+	copy(buf, b.buf)
+	b.buf = buf
+	b.cap = newCap
+	return nil
+}
+
+func makeSlice(n int64) (buf []byte, err error) {
+	defer func() {
+		if recover() != nil {
+			err = ErrOutOfMemory
+		}
+	}()
+	buf = make([]byte, n)
+	return
+}
+
+func (b *buffer) Write(p []byte) (int, error) {
+	n, err := b.WriteAt(p, b.cursor)
+	if err != nil {
+		return 0, err
+	}
+	b.cursor += int64(len(p))
+	if b.end < b.cursor {
+		b.end = b.cursor
+	}
+	return n, nil
+}
+
+func (b *buffer) WriteAt(p []byte, off int64) (int, error) {
```

Comment:
I would much prefer to revert to storing contents of files as they are downloaded in a temporary file when the ultimate destination is a `[]byte`, a buffer implementation (and one for safe concurrent access at that) will add much more complexity to the Ignition code base than I feel is necessary here.

Since you're the one who requested I switch to a custom buffer implementation in the first place, what are your thoughts @crawford ?